packages/paths/src/index.ts

```diff
@@ -0,0 +1,3 @@
+export * from './config'
```

Comment:
I think having the index file in this package is ok, but for the next breakout package, we need to make sure we don't have a index exporting everything... this is where our performance issues come from.

Reply:
As long as the package is small enough the index file isn't a problem though, is it?

If we don't have an index file, what do we use for "main" and "types" in package.json

Can we introduce tooling that checks package sizes instead of not allowing index.ts files?

Reply:
It depends on what you mean my "size" - the package maybe small today, but tomorrow I may add an import to graphql-codegen or some heavy module without realising, which will cause memory usage to spiral. 

Redwoodjs/testing/api has this problem for example, with the directive test utils. Clearly when we added these, we weren't trying to increase memory usage, but it happens over time if you are not careful, and have a root index.ts pattern.

You can have an index file, but don't export anything from it OR you don't need a main at all.

Reply:
> It depends on what you mean my "size" - the package maybe small today, but tomorrow I may add an import to graphql-codegen or some heavy module without realising, which will cause memory usage to spiral.

You're going to have to add that graphql-codegen export to _some_ file. If not index.ts it's going to be some other file, and now you've caused memory usage to spiral for whoever imports straight from there, perhaps without realizing. 

I still think tooling that measures package sizes and warns when they grow too big is a better solution to the problem.

Reply:
Some links etc here that we might find useful https://github.com/prisma/prisma/issues/10356

Reply:
Happy with more tooling, but our problem is _not_ bundle size - we don't bundle any of these. Our problem is memory usage and require stack size. 

If there's tooling that catches this it would be awesome!

Reply:
I'm not making a made up argument here.... the massive CLI speed boost and memory usage drop came from this optimisation + moving handlers to be loaded when requried (which is the same optimisation, applied elsewhere)

Reply:
Of course you're not making anything up 😄

But you could have made the exact same speed boosts and memory usage drops by splitting things out into separate packages (with index exports), right?

And it's equally easy to accidentally add a heavy import to `packages/internal/src/paths.ts` as to `packages/paths/src/index.ts`, right?

Not having to remember to do dist imports is a plus for separate packages with index exports
Being able to use `nx graph` for checking dependencies is a plus for separate packages

Reply:
As far as tooling goes, I wonder if we could do something with https://nodejs.org/api/process.html#process_process_memoryusage Like if we could measure while running the tutorial tests because then we would capture usage of both cli commands like generators etc and rw dev

Reply:
> And it's equally easy to accidentally add a heavy import to packages/internal/src/paths.ts as to packages/paths/src/index.ts, right?

Not at all... this is my point. Exactly the opposite. The index exports leans into accidentally doing this more often (which is exactly why its happened so many times in our packages). Most recently in cache.

When you are creating a new function, and adding it to a package, you're _not_ thinking about all the other places this package would be used, you are only thinking about your particular use case. 

If I added a heavy function to `packages/paths/bazinga` now _anything_  that uses packages/paths will become heavy, if I have index exports. Realistically is someone going to check memory usage in all the other cases? Ofcourse not. 

Sometimes the simplest solution is the best one, and in this case, using a lint rule as well as removing index exports. By not having index exports, we prevent this problem from happening in the first place.

Reply:
This is great Tobbe, yes education and communication more within the core team. But its simply not realistic to expect a non-core contributor to understand all these quirks when they are trying to add a piece of functionality. We need the code and the patterns in place so that its easy to contribute (and less easy to make mistakes). 

This is why removing the index is the only realistic option here. It's much harder for someone to make a mistake like this, when they notice that the only way you import things from a package is via subpaths (whether via `dist` or via package aliasing), and never allowed to export from index. Is there a specific reason you are so adamant about having an index file?  

Anyway... I don't think you'll be convinced here - this is just learnings from having to fix the same problems over and over again - so your call how you want to proceed, as long as you are willing to take the responsibility for fixing any perf/memory issues that popup!

Reply:
> Is there a specific reason you are so adamant about having an index file?

I mentioned a couple of things earlier. 

Not having to remember to do dist imports is a plus for separate packages with index exports
Being able to use nx graph for checking dependencies is a plus for separate packages

I realize now however that the first one is easily remedied by just expanding that lint rule we already have.

But the biggest reason is that it's "weird". 
If there's one thing I've learned it's that as soon as you try to do some thing "cleaver" you end up spending the rest of your days fighting the ecosystem and all tooling. Having an index file is what everyone else does, so it's going to have the best support. Like `nx graph` support as I mentioned. 
It's funny you mentioned contributors. I see it in a different way. "What is this weird /dist import stuff? I don't understand this so I better not try to contribute". Vs just getting feedback on a PR "please don't use this heavy dependency" or "please lazy load this heavy dependency" or whatever the correct solution might be. 

Also had these two points, but 1. I came up with these now, so not part of why I initially wanted to do index exports. And 2. we would get the same benefits doing smaller packages with /dist imports. So index vs /dist doesn't really matter here
 * Smaller packages (which doing index exports might force us towards) is going to make it easier to avoid circular dependencies.
 * More/smaller packages is also going to benefit build and test caching thus increasing build+test speed.

Reply:
This isn't really going to go anywhere so here's my peace offering! 

https://user-images.githubusercontent.com/1521877/203690060-8a658d08-59b1-424a-9458-aedde89938e3.mp4

Just keep in mind that these issues don't ever get caught at the time of the PR (so any tooling here watching the require stack would be cool) - I've dealt with similar issues about 10/15 times now. Whether that's deployment failures to perf issues - since about RW 0.13